src/NumSharp.Core/Operations/Elementwise/NDArray.AND.cs

```diff
@@ -1,36 +1,168 @@
+        private static NDArray<TResult> Apply<TElement, TResult>(NDArray lhs, TElement rhs, Func<TElement, TElement, TResult> projector)
+            where TElement : unmanaged
+            where TResult : unmanaged
+        {
+            if (lhs is null)
+                throw new ArgumentNullException(nameof(lhs));
+            if (projector is null)
+                throw new ArgumentNullException(nameof(projector));
+
+            var result = new NDArray(typeof(TResult), lhs.shape);
+            var storage = result.Storage.GetData<TResult>();
+
+            var lha = lhs.Storage.GetData<TElement>();
+
+            for (var i = 0; i < storage.Count; i++)
+                storage[i] = projector(lha[i], rhs);
+
+            return result.AsGeneric<TResult>();
+        }
+
+        /// <summary>
+        /// Aplies the (x, y) => z projection delegate to each element of <see cref="NDArray"/> and a scalar and converts it to <typeparamref name="TResult"/>.
+        /// </summary>
+        /// <typeparam name="TElement">The type of <see cref="NDArray"/> element.</typeparam>
+        /// <typeparam name="TResult">The type of resulting element.</typeparam>
+        /// <param name="lhs">The left <see cref="NDArray"/> to pass to projector.</param>
+        /// <param name="rhs">The right <typeparamref name="TElement"/> scalar value to pass to projector.</param>
+        /// <param name="projector">The projection delegate that handles the array elements.</param>
+        /// <returns>The resulting <see cref="NDArray{TDType}"/> that contains the projection result.</returns>
+        private static NDArray<TResult> Apply<TElement, TResult>(NDArray lhs, NDArray rhs, Func<TElement, TElement, TResult> projector)
+            where TElement : unmanaged
+            where TResult : unmanaged
         {
-            return null;
-            //var boolTensor = new NDArray(typeof(bool),lhs.shape);
-            //bool[] bools = boolTensor.Storage.GetData<bool>();
+            if (rhs is null)
+                throw new ArgumentNullException(nameof(rhs));
+            if (lhs is null)
+                throw new ArgumentNullException(nameof(lhs));
+            if (projector is null)
+                throw new ArgumentNullException(nameof(projector));
 
-            //bool[] np = lhs.Storage.GetData<bool>();
-            //bool[] obj = rhs.Storage.GetData<bool>();
+            var result = new NDArray(typeof(TResult), lhs.shape);
+            var storage = result.Storage.GetData<TResult>();
 
-            // for(int i = 0;i < bools.Length;i++)
-            //    bools[i] = np[i] && obj[i];
+            var lha = lhs.Storage.GetData<TElement>();
+            var rha = rhs.Storage.GetData<TElement>();
 
-            //return boolTensor.MakeGeneric<bool>();
+            for (var i = 0; i < storage.Count; i++)
+                storage[i] = projector(lha[i], rha[i]);
+
+            return result.AsGeneric<TResult>();
```

Comment:
This will fail if you pass collections that are not of the same size. You need to add support for broadcasting. I highly suggest you look for similar unary operators to see how its done there.

You got to cover the following shapes:
(1), (1)
(10), (1)
(10), (10)

(1,4), (4, 1)
(4, 4), (4, 4)
(1, 1), (1, 4)
(1,1,1), (1,2,2)
(1), (1, 4)
(4), (1, 4)

Reply:
Yeah, you're right, this is why I said that I'm not sure about the input shapes.

I went through some binary operations and found that they also use the result type inference based on the input argument type codes. The existing methods are implemented using switch copy-pasted operation bodies based on the left+right code combinations. This significantly enlarges code base  and reduce maintainability, so I decided to suggest architectural solution to that problem.

It's based on operation->operator concept controlled by dynamic indexers. The base operation body is a generic function that handles the type inference and validation, detects broadcasting and passes control to the operator, passed as a parameter.

The operation is basically a container for generic delegate that will be called when operation executes, The operation indexer is to be declared statically as it keeps thread-safe index for operations when they're called. There's a small performance drawback at first invocation cause it uses reflection, and caches the operation instance. The indexer keeps separate operation for every (left arg, right arg, return value) type tuple.

The operator is the same generic delegate container but for the operator function. The idea is that the body has direct access to the data storage and the mapped operation, knowing the actual generic types. The operator also knows all generic types and can handle the pointers in the most effective way as it can reduce data translations/calls to the minimum. There're more ways here to for further optimization herewith lambda function generators. It also makes possible not to rely on operation bodies when handling arrays and make the operator nature independent of the higher level logic.

I have added the suggested shape test cases to the NDArray & NDArray, so they work fine. I understand that this change is not final so I suggest you to review the architectural update while I'll be preparing the additional test coverage for the new code.